fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/mixin/ingredient/IngredientMixin.java

```diff
@@ -0,0 +1,125 @@
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.JsonHelper;
+
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredient;
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredientSerializer;
+import net.fabricmc.fabric.api.ingredient.v1.FabricIngredient;
+import net.fabricmc.fabric.impl.ingredient.CustomIngredientImpl;
+import net.fabricmc.fabric.impl.ingredient.builtin.OrIngredient;
+
+@Mixin(value = Ingredient.class, priority = 900) // overwrite before faux ingredient extension API injects in the method
+public class IngredientMixin implements FabricIngredient {
+	/**
+	 * Inject right when vanilla detected a json object and check for our custom key.
+	 */
+	@Inject(
+			at = @At(
+					value = "INVOKE",
+					target = "net/minecraft/recipe/Ingredient.entryFromJson (Lcom/google/gson/JsonObject;)Lnet/minecraft/recipe/Ingredient$Entry;",
+					ordinal = 0
+			),
+			method = "fromJson",
+			cancellable = true
+	)
+	private static void injectFromJson(JsonElement json, CallbackInfoReturnable<Ingredient> cir) {
+		JsonObject obj = json.getAsJsonObject();
+
+		if (obj.has(CustomIngredientImpl.TYPE_KEY)) {
+			Identifier id = new Identifier(JsonHelper.getString(obj, CustomIngredientImpl.TYPE_KEY));
+			CustomIngredientSerializer<?> serializer = CustomIngredientSerializer.get(id);
+
+			if (serializer != null) {
+				cir.setReturnValue(serializer.read(obj).toVanilla());
+			} else {
+				throw new IllegalArgumentException("Unknown custom ingredient type: " + id);
+			}
+		}
+	}
+
+	/**
+	 * Throw exception when someone attempts to use our custom key inside an array ingredient.
+	 * The {@link OrIngredient} should be used instead.
+	 */
+	@Inject(at = @At("HEAD"), method = "entryFromJson")
+	private static void injectEntryFromJson(JsonObject obj, CallbackInfoReturnable<?> cir) {
+		if (obj.has(CustomIngredientImpl.TYPE_KEY)) {
+			throw new IllegalArgumentException("Custom ingredient cannot be used inside an array ingredient. You can replace the array by a fabric:or ingredient.");
+		}
+	}
+
+	@Inject(
+			at = @At(
+					value = "INVOKE",
+					target = "net/minecraft/network/PacketByteBuf.writeCollection (Ljava/util/Collection;Lnet/minecraft/network/PacketByteBuf$PacketWriter;)V"
+			),
+			method = "write",
+			cancellable = true
+	)
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	private void injectWrite(PacketByteBuf buf, CallbackInfo ci) {
+		// Done here instead of writing the ingredient in the subclass for compat with faux ingredient extension API
+		CustomIngredient customIngredient = getCustomIngredient();
+
+		if (customIngredient != null) {
```

Comment:
You could check with networking api if client has required fabric version for better compatibility. Also it could be used to skip ingredients unsupported by client

Reply:
I can't skip ingredients that are not supported by the client due to how reloading works, but I can kick clients that don't have the ingredient API if the recipe packet contains custom ingredients. I'll try that.

Reply:
Because that method is written on networking thread, you can get the player, see Nucleoid's Packet Tweaker library

Reply:
Look at the way reload is implemented. It's using a single packet for all connected clients.

Reply:
Recipes are still written in separate bytebufs per player (packet only contains list of recipes), so that would still work (polymer works thanks to that for example)

Reply:
Ahhh I see, I'll have a look. Thanks!

Should I just skip invalid recipes, perhaps with a console warning, or disconnect the player?

Reply:
To be fair looking at it now, writing result of getMatchingStacks for client not supporting required type would be also a good alternative

Reply:
It might lead to ugly desyncs though, where the client thinks it can craft something but the server disagrees. Maybe we should write a barrier block with a warning in the name, Forge-style. :joy:

Reply:
Well, adding some random nbt (with hard to get with commands key/value) would resolve auto matching